src/Microsoft.CodeQuality.Analyzers.Exp/Core/FlowAnalysis/Analysis/NullAnalysis/NullAnalysis.NullDataFlowOperationVisitor.cs

```diff
@@ -14,23 +14,20 @@ internal partial class NullAnalysis : ForwardDataFlowAnalysis<NullAnalysisData,
         private sealed class NullDataFlowOperationVisitor : DataFlowOperationVisitor<NullAnalysisData, NullAbstractValue>
         {
             public NullDataFlowOperationVisitor(
-                AbstractDomain<NullAbstractValue> valueDomain,
+                NullAbstractValueDomain valueDomain,
                 INamedTypeSymbol containingTypeSymbol,
                 DataFlowAnalysisResult<PointsToAnalysis.PointsToBlockAnalysisResult, PointsToAnalysis.PointsToAbstractValue> pointsToAnalysisResultOpt)
                 : base(valueDomain, containingTypeSymbol, nullAnalysisResultOpt: null, pointsToAnalysisResultOpt: pointsToAnalysisResultOpt)
             {
             }
 
-            protected override NullAbstractValue UnknownOrMayBeValue => NullAbstractValue.MaybeNull;
+            protected override IEnumerable<AnalysisEntity> TrackedEntities => CurrentAnalysisData.Keys;
```

Comment:
does people need to care about concurrency? or it doesn't matter?

Reply:
Current design of the analyses is to create a separate instance of the analysis/visitor object to perform the analysis on a given CFG. The analysis itself operates sequentially on the flow graph and doesn't need to be thread-safe. The result produced by the analysis is pure immutable data which can be accessed concurrently and must be thread-safe.